test/test_pebble.py

```diff
@@ -94,7 +95,7 @@ def test_api_error(self):
 
     def test_change_error(self):
         change = pebble.Change(
-            id=pebble.ChangeID('1234'),
+            id=typing.cast(pebble.ChangeID, '1234'),
```

Comment:
What's the reason for this change? If I change it back to `pebble.ChangeID('1234')` locally, Pyright is fine with that.

Similar below.

Reply:
My bad, sorry. I had added `ChangeID('blah')` (or `TaskID`) in a few places, and then when reviewing the code I recalled a conversation in one of the other PRs about these (it was one of the Dicts in that case, I think) being more for declaring types than for actually creating objects. So I change them back but did that over-aggressively and changed some that already existed.

For clarity (mine and maybe I can incorporate this into some docs somewhere when working on those): what is the intention with these types - the `*ID` ones and also the `*Dict` ones (especially if we remove the leading underscore as discussed elsewhere here)?

Ignoring that you'd probably collapse these to a single line, which of these (if any) would you recommend for a charmer?

```python
task_id = pebble.TaskID('12345')
task = ops.pebble.Task(id=task_id, kind='start', ...)
```

```python
task_id: pebble.TaskID = '12345'  # pyright will complain about this
task = ops.pebble.Task(id=task_id, kind='start', ...)
```

```python
task_id = typing.cast(pebble.TaskID, '12345')
task = ops.pebble.Task(id=task_id, kind='start', ...)
```

And would your answer be the same for ChangeDict, WarningDict, ChangeID, etc?

Reply:
Yeah, the intention with these types is something devs using strongly-typed languages like Scala sometimes do: create specific named types so it's harder to accidentally pass (say) a TaskID string as a ChangeID, or a Name as an Email. *However*, creating named types just to differentiate kinds of strings isn't something that I've seen much in Python (or Go for that matter -- some do it but it's rare), so I was probably a bit overzealous when adding those types in the first place -- I don't think I'd do that now, but we're stuck with these. :-)

In any case, it's a bit different from TypedDict: that's just a normal dict with a specific shape/fields. So for those I'd probably use `d: ChangeDict = {"foo": 42}` or probably just `d = {"foo": 42}` (it just needs to fit the parameter types shape, not be that type).

But these are actual types (a subclass of str), so I think it makes most sense to instantiate them as `pebble.TaskID('123')` -- your first example.

In real charm code, however, you probably never have to say `pebble.ChangeID('123')`, as you'd probably get a change ID from `Client.get_changes` or similar (which returns it as a `ChangeID` type already), and then pass it directly to `get_change` or similar.

Reply:
Hmm, I was wrong about `TypedDict` -- they're not just a normal dict with shape checked, but a real type. If you say:

```python
        d = {
            'message': 'Look out...',
            'first-added': '2020-12-25T17:18:54.016273778+13:00',
            'last-added': '2021-01-26T17:01:02.12345+13:00',
            'expire-after': '1s',
            'repeat-after': '2s',
        }
        warning = pebble.Warning.from_dict(d)
```

Pyright gives this error:

```
Argument of type "dict[str, str]" cannot be assigned to parameter "d" of type "WarningDict" in function "from_dict"
```

However, type checkers *should* automatically infer the type if you pass it directly as a function argument. So this type checks fine:

```python
        warning = pebble.Warning.from_dict({
            'message': 'Look out...',
            'first-added': '2020-12-25T17:18:54.016273778+13:00',
            'last-added': '2021-01-26T17:01:02.12345+13:00',
            'expire-after': '1s',
            'repeat-after': '2s',
        })
```

Ah, the joys of Python type checking. See more at https://peps.python.org/pep-0589/#using-typeddict-types

I guess that kind of changes my opinion about `d = ChangeDict({"foo": 42})` -- that does seem reasonable from a typing point of view. But I still prefer to think of `ChangeDict` as "basically a normal dict with a shape", so IMO we should use:

* `d: ChangeDict = {'foo': 42}` style for typed dicts, and
* `task_id = pebble.TaskID('123')` style for those subclass types